sw/host/opentitanlib/src/transport/hyperdebug/spi.rs

```diff
@@ -0,0 +1,209 @@
+    max_chunk_size: usize,
+}
+
+const USB_SPI_PKT_ID_CMD_GET_USB_SPI_CONFIG: u16 = 0;
+const USB_SPI_PKT_ID_RSP_USB_SPI_CONFIG: u16 = 1;
+const USB_SPI_PKT_ID_CMD_TRANSFER_START: u16 = 2;
+const USB_SPI_PKT_ID_CMD_TRANSFER_CONTINUE: u16 = 3;
+//const USB_SPI_PKT_ID_CMD_RESTART_RESPONSE: u16 = 4;
+const USB_SPI_PKT_ID_RSP_TRANSFER_START: u16 = 5;
+const USB_SPI_PKT_ID_RSP_TRANSFER_CONTINUE: u16 = 6;
+
+//const USB_SPI_REQ_DISABLE: u8 = 1;
+const USB_SPI_REQ_ENABLE: u8 = 0;
+
+const USB_MAX_SIZE: usize = 64;
+const FULL_DUPLEX: usize = 65535;
+
+impl HyperdebugSpiTarget {
+    pub fn open(hyperdebug: &Hyperdebug, idx: u8) -> Result<Self> {
+        let mut usb_handle = hyperdebug.inner.usb_device.borrow_mut();
+        
+        // Tell HyperDebug to enable SPI bridge.
+        usb_handle.write_control(
+            rusb::request_type(Direction::Out, RequestType::Vendor, Recipient::Interface),
+            USB_SPI_REQ_ENABLE,
+            0 /* wValue */,
+            hyperdebug.spi_interface.interface as u16,
+            &mut [])?;
+
+        // Exclusively claim SPI interface, preparing for bulk transfers.
+        usb_handle.claim_interface(hyperdebug.spi_interface.interface)?;
+
+        // Initial bulk request/response to query capabilities.
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_GET_USB_SPI_CONFIG.to_le_bytes());
+        usb_handle.write_bulk(hyperdebug.spi_interface.out_endpoint, &buf[0..2])?;
+        let rc = usb_handle.read_bulk(hyperdebug.spi_interface.in_endpoint, &mut buf)?;
+        ensure!(rc == 8, Error::CommunicationError);
+        ensure!(buf[0..2] == USB_SPI_PKT_ID_RSP_USB_SPI_CONFIG.to_le_bytes(),
+                Error::CommunicationError);
+        let max_write_chunk = u16::from_le_bytes(buf[2..4].try_into()?);
+        let max_read_chunk = u16::from_le_bytes(buf[4..6].try_into()?);
+        let feature_bitmap = u16::from_le_bytes(buf[6..8].try_into()?);
+        // Verify that interface supports concurrent read/write.
+        ensure!((feature_bitmap & 0x0001) != 0, Error::CommunicationError);
+        
+        Ok(Self {
+            inner: Rc::clone(&hyperdebug.inner),
+            interface: hyperdebug.spi_interface,
+            _target_idx: idx,
+            max_chunk_size: std::cmp::min(max_write_chunk, max_read_chunk) as usize,
+        })
+    }
+
+    /// Transmit data for a single SPI operation, using one or more USB packets.
+    fn transmit(&self, wbuf: &[u8], rbuf_len: usize) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_TRANSFER_START.to_le_bytes());
+        buf[2..4].clone_from_slice(&(wbuf.len() as u16).to_le_bytes());
+        buf[4..6].clone_from_slice(&(rbuf_len as u16).to_le_bytes());
+        let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+        buf[6..6 + databytes].clone_from_slice(&wbuf[0..databytes]);
```

Comment:
This looks like `buf` really wants to be a struct with some named fields.  You might take a look at the `zerocopy::AsBytes` trait (used by `bootstrap/primitive.rs`) to write this as a struct and then hand it off to `usb_write_bulk` as a byte buffer.

Reply:
AsBytes would take care of the sending of datagrams, but decoding received datagrams would still have to be done differently (as_bytes() on return an immutable reference).  I saw that a FromBytes trait exists, but I am not sure if it can be used without resorting to "unsafe".
Also, if we ever compile OpenTitanTool on a big-endian architecture, I am afraid that AsBytes would fail to provide the little endian encoding required in this protocol.

I would prefer to keep encoding and decoding somewhat similar.  Do you have a example of how to handle decoding of byte arrays into a struct in Rust?

Reply:
The `Frame` structures in the bootstrap implementation are already tied to a little-endian architecture.

As for writing into a struct from binary, data, I think you can derive `FromBytes` and then use `AsBytes::as_bytes_mut()` to get a mutable buffer into which you can write data.

https://docs.rs/zerocopy/0.6.1/zerocopy/trait.AsBytes.html#method.as_bytes_mut

Reply:
It may be true that big-endian architectures are a thing of the past. In my first real job, our software ran on AIX, Solaris, HP Unix, and all sorts of strange machines, so keeping track of when something was in machine endian-ness and when it was in a particular wire format with fixed endian-ness was imperative.

I have changed the code to use AsBytes/FromBytes.